state/src/account.rs

```diff
@@ -0,0 +1,401 @@
+        StateObject {
+            balance,
+            nonce,
+            storage_root: SHA3_NULL_RLP,
+            code_hash: SHA3_EMPTY,
+            code: vec![],
+            code_size: 0,
+            code_state: CodeState::Clean,
+            storage_changes: HashMap::new(),
+        }
+    }
+
+    /// Create a new account from rlp bytes.
+    /// Note: make sure you use `read_code` after this.
+    pub fn from_rlp(data: &[u8]) -> Result<StateObject, Error> {
+        let account: Account = rlp::decode(data)?;
+        Ok(account.into())
+    }
+
+    pub fn rlp(&self) -> Vec<u8> {
+        rlp::encode(&self.account())
+    }
+
+    pub fn account(&self) -> Account {
+        Account {
+            balance: self.balance,
+            nonce: self.nonce,
+            storage_root: self.storage_root,
+            code_hash: self.code_hash,
+        }
+    }
+
+    pub fn init_code(&mut self, code: Vec<u8>) {
+        self.code = code.clone();
+        self.code_hash = From::from(&Sha3_256::digest(&code)[..]);
+        self.code_size = code.len();
+        self.code_state = CodeState::Dirty;
+    }
+
+    pub fn read_code<B: DB>(&mut self, db: &mut B) -> Result<Vec<u8>, Error> {
+        if self.code_hash == SHA3_EMPTY {
+            return Ok(vec![]);
+        }
+        if !self.code.is_empty() {
+            return Ok(self.code.clone());
+        }
+        let c = db
+            .get(&self.code_hash)
+            .or_else(|e| Err(Error::DB(format!("{}", e))))?
+            .unwrap_or_else(|| vec![]);
+        self.code = c.clone();
+        self.code_size = c.len();
+        self.code_state = CodeState::Clean;
+        Ok(c)
+    }
+
+    pub fn balance(&self) -> U256 {
+        self.balance
+    }
+
+    pub fn nonce(&self) -> U256 {
+        self.nonce
+    }
+
+    pub fn code(&self) -> Option<Vec<u8>> {
+        if self.code.is_empty() {
+            return None;
+        }
+        Some(self.code.clone())
+    }
+
+    pub fn code_hash(&self) -> H256 {
+        self.code_hash
+    }
+
+    pub fn code_size(&self) -> usize {
+        self.code_size
+    }
+
+    pub fn inc_nonce(&mut self) {
+        self.nonce += U256::from(1u8);
+    }
+
+    pub fn add_balance(&mut self, x: U256) {
+        let (a, b) = self.balance.overflowing_add(x);
+        // overflow is not allowed at state_object.
+        assert_eq!(b, false);
+        self.balance = a;
+    }
+
+    pub fn sub_balance(&mut self, x: U256) {
+        let (a, b) = self.balance.overflowing_sub(x);
+        assert_eq!(b, false);
+        self.balance = a;
+    }
+
+    pub fn set_storage(&mut self, key: H256, value: H256) {
+        self.storage_changes.insert(key, value);
+    }
+
+    pub fn get_storage_at_backend<B: DB>(
+        &mut self,
+        db: &mut B,
+        key: &H256,
+    ) -> Result<Option<H256>, Error> {
+        let trie = PatriciaTrie::from(db, RLPNodeCodec::default(), &self.storage_root.0)
+            .or_else(|e| Err(Error::DB(format!("{}", e))))?;
```

Comment:
Need to create a new `trie` each time? In this way, you need to regenerate the full path every time you query,  the performance is very poor.

Reply:
Emmm...  I will spend some time considering about it.


Parity reconsturct the trie every time when `storage_at()` called, See


https://github.com/paritytech/parity-ethereum/blob/master/ethcore/src/state/account.rs#L224

https://github.com/paritytech/parity-ethereum/blob/master/ethcore/src/state/account.rs#L255

Reply:
When querying an uncached key, if you can reuse the previous path, you can reduce the consumption of many database IO and recovery `node`.

However, this will use more memory and the number of `node` that need to be calculated per `commit` will increase, because the `hash node` is restored from the database.

So this optimization requires serious thinking.

Reply:
Have a look at: https://github.com/cryptape/cita-trie/issues/16